secret-rotator/svckey/svckey.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package svckey
+
+// package svckey implements the provisioning and deactivation of gcloud service account keys
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"google.golang.org/api/iam/v1"
+	"k8s.io/klog"
+	"strings"
+)
+
+type ServiceAccountKeySpec struct {
+	Project        string `yaml:"project"`
+	ServiceAccount string `yaml:"serviceAccount"`
+}
+
+func (svc ServiceAccountKeySpec) String() string {
+	return fmt.Sprintf("projects/%s/serviceAccounts/%s", svc.Project, svc.ServiceAccount)
+}
+
+// ServiceAccountKeySpec.Type() is used to obtain the provisioner of the ServiceAccountKey
+func (svc ServiceAccountKeySpec) Type() string {
+	return "serviceAccountKey"
+}
+
+// ServiceAccountKeySpec.Labels() is used to obtain the labels needed for the provisioner of the ServiceAccountKey
+func (svc ServiceAccountKeySpec) Labels() map[string]string {
+	return map[string]string{
+		"project":         svc.Project,
+		"service-account": svc.ServiceAccount,
+	}
+}
+
+type Provisioner struct {
+	Service *iam.Service
+}
+
+func NewProvisioner() (*Provisioner, error) {
+	ctx := context.Background()
+
+	service, err := iam.NewService(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Provisioner{service}, nil
+}
+
+// TODO: add project and service-account fields to metadata when registered
+
+// CreateNew provisions a new service account key,
+// returns the key-id and private-key data of the created key if successful,
+// otherwise returns error
+func (p *Provisioner) CreateNew(labels map[string]string) (string, []byte, error) {
+	name := fmt.Sprintf("projects/%s/serviceAccounts/%s@%s.iam.gserviceaccount.com", labels["project"], labels["service-account"], labels["project"])
```

Comment:
I think the drawbacks are that for other types of provisioner (e.g. ssh key provisioner), `project, serviceAccount string` wouldn't be the right things to pass in, and might also require different number of fields.
I was trying to come up with an mostly general `Provisioner` interface. Although we can use variadic function parameters for `Provisioner.CreateNew(para ...string)`, that still implicitly requires those labels to exist.

Answering the first question, these labels should exist since they are attach to the input `labels` by https://github.com/kubernetes-sigs/k8s-gsm-tools/blob/15c29995e3dedf5e1f15e3667f65c7871204d7f3/secret-rotator/svckey/svckey.go#L42 and https://github.com/kubernetes-sigs/k8s-gsm-tools/blob/15c29995e3dedf5e1f15e3667f65c7871204d7f3/secret-rotator/rotator/rotator.go#L104.

There's one thing missing though is the Validate function for `ServiceAccountKeySpec` to ensure `Project` and `ServiceAccount` exist in the configMap.